### ISSUE TITLE
Assignment of incompatible function types allowed

### ISSUE THREAD

**Issue by webstrand** [Microsoft/TypeScript#29120]
**TypeScript Version:**  3.3.0-dev.20181221

**Search Terms:**
Assignment of functions with incompatible prototypes
incompatible parameters 

**Code**

```ts
interface Parent { shared: string }
interface Child extends Parent { child_only: string }

interface UnrelatedWide { shared: string }
interface UnrelatedNarrow { shared: string, child_only: string }

// Expected: Parent is not assignable to Child
var inval1: (a: Parent) => string = (a: Child) => a.child_only;

// Expected: string is not assignable to "???"
var inval2: (a: string) => string = (a: "???") => "";

// Expected: UnrelatedWide is not assignable to UnrelatedNarrow
var inval3: (a: UnrelatedWide) => string = (a: UnrelatedNarrow) => a.child_only;
```

**Expected behavior:**
For ```inval1``` I expected to receive a Compiler Error due to incompatible assignment from ```Parent``` to ```Child```, because ```Child``` expects to have the ```child_only``` property which Parent does not provide.

For ```inval2``` I expected to receive a Compiler Error due to incompatible assignment from string to the narrower type ```"???"```.

For ```inval3``` the expected behaviour is the same as ```inval1```.

**Actual behavior:**
Passes type check with no errors.

[**Playground Link**](https://www.typescriptlang.org/play/#src=type%20CONST%20%3D%20%22const%22%3B%0D%0A%0D%0Ainterface%20Parent%20%7B%20shared%3A%20string%20%7D%0D%0Ainterface%20Child%20extends%20Parent%20%7B%20child_only%3A%20string%20%7D%0D%0A%0D%0Ainterface%20UnrelatedWide%20%7B%20shared%3A%20string%20%7D%0D%0Ainterface%20UnrelatedNarrow%20%7B%20shared%3A%20string%2C%20child_only%3A%20string%20%7D%0D%0A%0D%0A%2F%2F%20Expected%3A%20Parent%20is%20not%20assignable%20to%20Child%0D%0Avar%20inval1%3A%20(a%3A%20Parent)%20%3D%3E%20string%20%3D%20(a%3A%20Child)%20%3D%3E%20a.child_only%3B%0D%0A%0D%0A%2F%2F%20Expected%3A%20string%20is%20not%20assignable%20to%20%22%3F%3F%3F%22%0D%0Avar%20inval2%3A%20(a%3A%20string)%20%3D%3E%20string%20%3D%20(a%3A%20%22%3F%3F%3F%22)%20%3D%3E%20%22%22%3B%0D%0A%0D%0A%2F%2F%20Expected%3A%20UnrelatedWide%20is%20not%20assignable%20to%20UnrelatedNarrow%0D%0Avar%20inval3%3A%20(a%3A%20UnrelatedWide)%20%3D%3E%20string%20%3D%20(a%3A%20UnrelatedNarrow)%20%3D%3E%20a.child_only%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%2F%2F%20Maybe%20related%3F%0D%0Atype%20Callable%20%3D%20(a%3A%20any%2C%20...args%3A%20any%5B%5D)%20%3D%3E%20any%3B%0D%0Atype%20ReplaceFirstArg%3CU%20extends%20Callable%2C%20V%3E%20%3D%20U%20extends%20(a%3A%20infer%20P%2C%20...args%3A%20infer%20Q)%20%3D%3E%20infer%20R%20%3F%20(a%3A%20P%20extends%20V%20%3F%20P%20%3A%20V%2C%20...args%3A%20Q)%20%3D%3E%20R%20%3A%20never%3B%0D%0Adeclare%20function%20test%3CU%20extends%20Callable%2C%20V%3E(fn%3A%20ReplaceFirstArg%3CU%2C%20V%3E%2C%20a%3A%20V)%3A%20void%3B%0D%0A%0D%0Atest(function%20(a%3A%20%22foo%22)%20%7B%20return%20a%20%2B%20%22%22%20%7D%2C%20%7B%7D)%3B%20)

**Related Issues:** 
```ts
type Callable = (a: any, ...args: any[]) => any;
type ReplaceFirstArg<U extends Callable, V> = U extends (a: infer P, ...args: infer Q) => infer R ? (a: P extends V ? P : V, ...args: Q) => R : never;
declare function test<U extends Callable, V>(fn: ReplaceFirstArg<U, V>, a: V): void;

// Expected: {} is not assignable to "foo"
test(function (a: "foo") { return "" }, {}); 
```
Might be related? I was trying to figure out why incompatible assignment is permitted through ```test``` when I discovered the above issue.

**Comment by jack-williams** [comment]
Do you have `--strictFunctionTypes` on?

**Comment by webstrand** [comment]
That's the issue: My project has --strictFunctionTypes enabled, but the playground–where I was debugging the related issue–does not enable that flag by default.
